File: java/connect/src/main/java/Connect.java
import java.util.ArrayDeque;
import java.util.Deque;

class Connect {

    private final char[][] grid;
    private final int height;
    private final int width;
    private static final char PLAYER_X = 'X';
    private static final char PLAYER_O = 'O';
    private static final char EMPTY = '.';

    public Connect(String[] board) {
        // Determine dimensions and parse the board, removing spaces
        height = board.length;
        if (height == 0) {
            width = 0;
            grid = new char[0][0];
            return;
        }
        // Calculate width based on the first row after removing spaces
        width = board[0].replace(" ", "").length();
        grid = new char[height][width];

        for (int i = 0; i < height; i++) {
            String row = board[i].replace(" ", "");
            for (int j = 0; j < width; j++) {
                grid[i][j] = row.charAt(j);
            }
        }
    }

    public Winner computeWinner() {
        if (height == 0 || width == 0) {
            return Winner.NONE;
        }

        if (checkWin(PLAYER_X)) {
            return Winner.PLAYER_X;
        }
        if (checkWin(PLAYER_O)) {
            return Winner.PLAYER_O;
        }
        return Winner.NONE;
    }

    private boolean checkWin(char player) {
        boolean[][] visited = new boolean[height][width];
        Deque<int[]> stack = new ArrayDeque<>();

        // Initialize stack based on player
        if (player == PLAYER_X) {
            // Start DFS from the left edge (first column)
            for (int i = 0; i < height; i++) {
                if (grid[i][0] == player) {
                    stack.push(new int[]{i, 0});
                    visited[i][0] = true;
                }
            }
        } else { // PLAYER_O
            // Start DFS from the top edge (first row)
            for (int j = 0; j < width; j++) {
                if (grid[0][j] == player) {
                    stack.push(new int[]{0, j});
                    visited[0][j] = true;
                }
            }
        }

        // Define the 6 possible neighbor offsets for hexagonal grid
        int[][] neighbors = {
            {-1, 0}, {-1, 1}, {0, -1}, {0, 1}, {1, -1}, {1, 0}
        };

        while (!stack.isEmpty()) {
            int[] current = stack.pop();
            int r = current[0];
            int c = current[1];

            // Check win condition based on player
            if (player == PLAYER_X && c == width - 1) {
                return true; // Reached the right edge
            }
            if (player == PLAYER_O && r == height - 1) {
                return true; // Reached the bottom edge
            }

            // Explore neighbors
            for (int[] offset : neighbors) {
                int nr = r + offset[0];
                int nc = c + offset[1];

                // Check bounds and if it's a valid move for the player
                if (nr >= 0 && nr < height && nc >= 0 && nc < width &&
                    grid[nr][nc] == player && !visited[nr][nc]) {
                    visited[nr][nc] = true;
                    stack.push(new int[]{nr, nc});
                }
            }
        }

        return false; // No winning path found
    }
}

File: javascript/connect/connect.js
//
// This is only a SKELETON file for the 'Connect' exercise. It's been provided as a
// convenience to get you started writing code faster.
//

export class Board {
  constructor(input) {
    this.size = input.length;
    this.board = input.map(row => row.trim().split(' '));
  }

  winner() {
    if (this.hasPath('X')) return 'X';
    if (this.hasPath('O')) return 'O';
    return '';
  }

  hasPath(player) {
    const visited = new Set();
    const stack = [];

    if (player === 'X') {
      // Start from all 'X' on the left edge
      for (let row = 0; row < this.size; row++) {
        if (this.board[row][0] === 'X') {
          stack.push([row, 0]);
          visited.add(`${row},0`);
        }
      }
    } else {
      // Start from all 'O' on the top edge
      for (let col = 0; col < this.board[0].length; col++) {
        if (this.board[0][col] === 'O') {
          stack.push([0, col]);
          visited.add(`0,${col}`);
        }
      }
    }

    const directions = [
      [-1, 0],  // up
      [-1, 1],  // up-right
      [0, -1],  // left
      [0, 1],   // right
      [1, -1],  // down-left
      [1, 0],   // down
    ];

    while (stack.length > 0) {
      const [r, c] = stack.pop();

      if (player === 'X' && c === this.board[r].length - 1) {
        return true;
      }
      if (player === 'O' && r === this.size - 1) {
        return true;
      }

      for (const [dr, dc] of directions) {
        const nr = r + dr;
        const nc = c + dc;
        if (
          nr >= 0 && nr < this.size &&
          nc >= 0 && nc < this.board[nr].length &&
          this.board[nr][nc] === player &&
          !visited.has(`${nr},${nc}`)
        ) {
          visited.add(`${nr},${nc}`);
          stack.push([nr, nc]);
        }
      }
    }

    return false;
  }
}
